File: lowestAncestor/src/LCADAG.java
import java.util.*;



public class LCADAG {

	 int totalVertices, totalEdges;
	 int[] indegree;
	 boolean[] visited, cycle;
	 boolean cycleExists;
	 LinkedList<Integer> graph[];

	public LCADAG(int v) {
		totalVertices = v;
		totalEdges = 0;
		graph = new LinkedList[v];
		for(int i = 0; i < v; i++)
		{
			graph[i] = new LinkedList<Integer>();
		}
		indegree = new int[v];
		visited = new boolean[v];
		cycle = new boolean[v];
		cycleExists=false;
	}


	public String addEdge(int node1, int node2) {// add edge from node1 to node2
		if (isValidVertex(node1) && isValidVertex(node2)) {
			this.totalEdges++;
			this.graph[node1].add(node2);
			this.indegree[node2]++;
			return "Edge Added";
		} else {
			System.out.println("Invalid vertex entered");
			return "Invalid vertex entered";
		}
	}

	public boolean isValidVertex(int v) {
		if (v < 0 || v >= this.totalVertices) {
			return false;
		} else {
			return true;
		}
	}

	public boolean checkCycle(int v) {
		this.visited[v] = true;
		this.cycle[v] = true;
		for (int i : this.graph[v]) {
			if (!this.visited[i]) {
				this.checkCycle(i);

			} else if (this.cycle[i]) {
				cycleExists=true;
				return true;

			}
		}
		this.cycle[v] = false;
		return false;
	}

	public int findLCA(int v, int w) {
		this.checkCycle(0);
		if (this.cycleExists) 
		{
			return -1;
		} else if (!this.isValidVertex(v) || !this.isValidVertex(w)) {
			
			return -1;
		}
		boolean found = false;
		LCADAG reversed = reverse();
		ArrayList<Integer> a1 = reversed.BFS(v);
		ArrayList<Integer> a2 = reversed.BFS(w);
		ArrayList<Integer> commonAncestors = new ArrayList<Integer>();


		for (int i = 0; i < a1.size(); i++) {
			for (int j = 0; j < a2.size(); j++) {
				System.out.println("a1 is " + a1.get(i));
				System.out.println("a2 is " + a2.get(j));

				if (a1.get(i) == a2.get(j)) {
					commonAncestors.add(a1.get(i));
					found = true;
				}
			}
		}

		if (found) {
			System.out.println("lca between " + v + " and " + w + " is " + commonAncestors.get(0));
			return commonAncestors.get(0);
		} else {
			System.out.println("lca between " + v + " and " + w + " is -1");
			return -1; 
		}
	}

	
	public ArrayList<Integer> BFS(int s) {
		LinkedList<Integer> queue = new LinkedList<Integer>();
		boolean visited[] = new boolean[this.totalVertices]; 
		ArrayList<Integer> order = new ArrayList<Integer>();
		
														
		

		visited[s] = true;
		queue.add(s);

		while (queue.size() != 0) {
			s = queue.poll(); 
			order.add(s);
			System.out.println("queue size is " + order.size());

			
			Iterator<Integer> i = this.graph[s].listIterator();

			while (i.hasNext()) {
				int n = i.next();
				if (!visited[n]) {
					visited[n] = true;
					queue.add(n);
				}
			}
		}
		return order;
	}

	
	public LCADAG reverse() {
		LCADAG reversed = new LCADAG(totalVertices);
		for (int v = 0; v < this.totalVertices; v++) {
			for (int i : this.graph[v]) {
				reversed.addEdge(i, v);
			}
		}
		return reversed;
	}
}

File: lowestAncestorC++/src/LCADAG.cpp
#include "LCADAG.h"

using namespace std; 


string newEdge(Graph *graph, int node1, int node2){
	
	if (isValidVertex(graph, node1) && isValidVertex(graph, node2)) {
			graph->totalEdges++;
			(graph->graph[node1]).push_back(node2);
			(graph->indegree[node2])++;
			return "Edge Added";
		} else {
			cout<<"invalid vertex entered";
			return "Invalid vertex entered";
		}
	
}

bool isValidVertex(Graph *graph, int v){
	if (v < 0 || v >= graph->totalVertices) {
			return false;
		} else {
			return true;
		}
	
	 
}

bool checkCycle(Graph *graph, int v) {
		graph->visited[v] = true;
		graph->cycle[v] = true;
		for (int i : graph->graph[v]) {
			if (!graph->visited[i]) {
				checkCycle(graph,i);
				
			} else if (graph->cycle[i]) {
				graph->cycleExists=true;

				return true;
			}
		}
		graph->cycle[v] = false;
		return false;
	}
	
int findLCA(Graph *graph, int v, int w) {
		checkCycle(graph,0);
		if (graph->cycleExists) 
		{
			return -1;
		} else if (!isValidVertex(graph,v) || !isValidVertex(graph,w)) {
			
			return -1;
		}
		bool found = false;
		Graph *reversed = reverse(graph);
		list<int> a1 = bfs(reversed, v);
		list<int> a2 = bfs(reversed, w);
		list<int> commonAncestors;
	    list<int>::iterator a1it,a2it,cait; 
		
		
		for (int i = 0; i<a1.size(); i++) {
			a1it = a1.begin();
			advance(a1it, i);
			for (int j = 0; j<a2.size(); j++) {

				  
				  a2it = a2.begin();
				advance(a2it, j);
		
				if(*a1it==*a2it){
					commonAncestors.push_back(*a1it);
					found = true;
				}
			}
		}

		if (found) {
			cait = commonAncestors.begin();
			return *cait;
		} else {
			return -1; 
		}
}



list<int> bfs(Graph *graph, int s){
	int v = graph->totalVertices;
	
	list<int> queue,order;
	bool *visited = new bool[v]; 
	for (int i=0; i<v; i++){
		visited[i]=false;
	}
   
  
	visited[s] = true;
	queue.push_back(s); 
  
   
    list<int>::iterator i; 
  
    while(!queue.empty()) 
    { 
        s = queue.front(); 
        order.push_back(s);
        queue.pop_front(); 
	
        for (int j=0; j<graph->graph[s].size(); j++)
			 
		
        { 
			i = graph->graph[s].begin();
			advance(i,j);
			
            if (!visited[*i]) 
            { 
				
                visited[*i] = true; 
                queue.push_back(*i); 
            } 
        } 
    }
    return order;
}

Graph *reverse(Graph *graph) {
		Graph *reversed = newGraph(graph->totalVertices);
		for (int v = 0; v < graph->totalVertices; v++) {
			for (int i : graph->graph[v]) {
				newEdge(reversed, i, v);
			}
		}
		return reversed;
}
